### ISSUE TITLE
Proposal: use of requireNamespace(QuietStop) instead of require to avoid search path pollution

### ISSUE THREAD

**Issue by lepennec** [topepo/caret#603]
Dear all,

I really like the _caret_ package and teach it as the _right way_ to do learn and predict in R. I'm however always annoyed by the fact that so many packages are attached along the way, polluting the search path. I was always wondering whether something can be done in such a complex meta-package setting... and I may have a solution.

Roughly speaking, it seems that using __requireNamespace__ and explicit package name scheme (::) in _caret_ is (almost) sufficient to get rid of this issue. I've made a fork [lepennec/caret](https://github.com/lepennec/caret) that replace all instances of __require__ by __requireNamespace__ in the package and modify a few packages (the ones required to run the tests and the 6 llast ones in alphabetic order). The only drawback, and it's a big one, is that _foreach_ and _doParallel_ have been made to use __library__ and not __requireNamespace__... so that the previous modifications are not sufficient. I've made a fork of _foreach_ [lepennec/foreach](https://github.com/lepennec/foreach) that add a new parameter .namespaces which plays the same role than .packages but use __requireNamespace__ instead of __library__. I've used it with success in the fork. Note that the same modification can easily be done in _doParallel_. 

To summarize, it seems that the search path pollution issue can be solved with the help of __requireNamespace__ but that it requires some modifications of _foreach_ and _doParallel_... If you are interested I can keep on the work..

Yours,

Erwan

**Comment by lepennec** [comment]
Another possibility is to move the requireNamespace call in caret instead of foreach... This is probably slightly slower but does not require any modification of the foreach family. This is the current solution in my fork...

**Comment by topepo** [comment]
> so many packages are attached

Do you mean by package in the Description file or the packages used by the models? 

It's a good idea and I'll take a look at the fork

**Comment by lepennec** [comment]
I mean the packages used by the models...

**Comment by topepo** [comment]
All of the code for the models in the [files directory](https://github.com/topepo/caret/tree/master/models/files) would need to be called by namespace. For example, [`AdaBag`](https://github.com/topepo/caret/blob/master/models/files/AdaBag.R) uses `rpart.control`, `bagging`, and a `predict` method. Otherwise the functions can't be found. 

This is a good example since `rpart` isn't in the `library` element (since it would be loaded when all the others are loaded). 

There are a set of [regression test files](https://github.com/topepo/caret/tree/master/RegressionTests/Code) that can be used to make sure that the models are running fine.

**Comment by lepennec** [comment]
Indeed ! I was waiting for your approval to start the tedious work of modifying every model... So far I've been able to make the proposed approach works for every model I have tried.  The tests would definitely be useful to verify that every model has been corrected.

**Comment by topepo** [comment]
As long as the tests pass, I think that this is a great idea. Let me know if you have any questions.

**Comment by lepennec** [comment]
I'm still working on it. I have now alreadu 222 models that pass the tests but I'm fighting for the last 18...
GFS.GCCL.R
GFS.Thrift.R
glmnet_h2o.R
lssvmRadial.R
M5.R
M5Rules.R
monmlp.R
ordinalNet.R
pcaNNet.R
penalized.R
plsRglm.R
preprocess_back.R
rfe_treebag.R
rrlda.R
sbf_train.R
sddaLDA.R
sddaQDA.R
SLAVE.R
spikeslab.R
I'm still working on them. I will not be able to make it for next week... but will do my best to propose a pull before the next release.

**Comment by topepo** [comment]
Not a problem. You can exclude the `sdda` models since I've taken them out of the package library. You might want to check your list against the ones that you find using 
```r
names(getModelInfo())
```
Thanks

**Comment by topepo** [comment]
I'm going to close this. Go ahead and send a PR when you think that you want me to take a look.

**Comment by lepennec** [comment]
OK
